lib/AstOfLlbc.ml

```diff
@@ -1025,57 +1026,68 @@ let rec expression_of_fn_ptr env depth (fn_ptr: C.fn_ptr) =
-              } in
-              let fn_ptr = fst3 (expression_of_fn_ptr env (depth ^ "  ") fn_ptr) in
-              fn_ptr
-            ) (trait_impl.required_methods @ trait_impl.provided_methods)
-
-        | Clause clause_id ->
-            (* Caller it itself polymorphic and refers to one of its own clauses to synthesize
-               the clause arguments at call-site. *)
-            List.rev (Krml.KList.filter_mapi (fun i (var, t) ->
-              match var with
-              | TraitClauseMethod (clause_id', _, _) when clause_id = clause_id' ->
-                  Some K.(with_type t (EBound i))
-              | _ ->
-                  None
-            ) env.binders)
-
-        | ParentClause (_instance_id, decl_id, clause_id) ->
-            let trait_decl = env.get_nth_trait_decl decl_id in
-            let name = string_of_name env trait_decl.name in
-            let clause_id = C.TraitClauseId.to_int clause_id in
-            let parent_clause = List.nth trait_decl.parent_clauses clause_id in
-            let parent_clause_decl = env.get_nth_trait_decl parent_clause.trait_id in
-            let parent_name = string_of_name env parent_clause_decl.name in
-            Krml.KPrint.bprintf "looking up parent clause #%d of decl=%s = %s\n" clause_id name
-              parent_name;
-            if List.mem parent_name blocklisted_trait_decls then
+          match trait_ref.trait_id with
+          | _ when List.mem name blocklisted_trait_decls ->
+              (* Trait not supported -- don't synthesize arguments *)
               []
-            else
-              failwith ("Don't know how to resolve trait_ref " ^ C.show_trait_ref trait_ref)
 
-        | _ ->
-            failwith ("Don't know how to resolve trait_ref " ^ C.show_trait_ref trait_ref)
+          | TraitImpl impl_id ->
+              (* Call-site has resolved trait clauses into a concrete trait implementation. *)
+              let trait_impl: C.trait_impl = env.get_nth_trait_impl impl_id in
+
+              (* This must be in agreement, and in the same order as build_trait_clause_mapping *)
+              List.map (fun (item_name, decl_id) ->
+                let fn_ptr: C.fn_ptr = {
+                  func = TraitMethod (trait_ref, item_name, decl_id);
+                  generics = Charon.TypesUtils.empty_generic_args
+                } in
+                let fn_ptr = fst3 (expression_of_fn_ptr env (depth ^ "  ") fn_ptr) in
+                fn_ptr
+              ) (trait_impl.required_methods @ trait_impl.provided_methods) @
+              build_trait_ref_mapping ("  " ^ depth) trait_ref.generics.trait_refs
+
+          | Clause _ as clause_id ->
+              (* Caller it itself polymorphic and refers to one of its own clauses to synthesize
+                 the clause arguments at call-site. We must pass whatever is relevant for this
+                 clause, *transitively* (this means all the reachable parents). *)
+              let rec relevant = function
```

Comment:
insofar as I understand, when the caller is clause-polymorphic, and the callee is clause-polymorphic too, then I only see a single `Clause _` reference at call-site, and it's up to me to remember that this trait bound requires parent bounds too, and materialize method references for the parent, too

Reply:
Mm yes, charon models a clause as "containing" its parent clauses, so we don't mention the parent clauses at call site. I think you should never think of a clause without its parents, so there should be no need for you to explore the parents. What are you trying to do here exactly?

Reply:
I think it's fine, yes